src/main/src/main/java/org/geoserver/ows/CiteComplianceHack.java

```diff
@@ -36,12 +44,66 @@ public boolean preHandle(
             throws Exception {
         if (handler instanceof Dispatcher) {
             Dispatcher dispatcher = (Dispatcher) handler;
-            dispatcher.setCiteCompliant(getInfo().isCiteCompliant());
+            Service service = findService(dispatcher, request, response);
+            if (service != null
+                    && (service.getId().equalsIgnoreCase(getInfo().getId())
+                            || service.getId().equalsIgnoreCase(getInfo().getName()))) {
+                dispatcher.setCiteCompliant(getInfo().isCiteCompliant());
+            }
         }
 
         return true;
     }
 
+    private Service findService(
+            Dispatcher dispatcher, HttpServletRequest request, HttpServletResponse response) {
+        // create a new request instance
+        Request req = new Request();
+
+        // set request / response
+        req.setHttpRequest(request);
+        req.setHttpResponse(response);
+
+        // find the service
+        try {
+            return dispatcher.service(req);
+        } catch (Throwable t) {
+            LOGGER.log(Level.FINE, "Exception while looking for the 'Service' from the request", t);
+            // load from teh context
+            try {
+                UriComponentsBuilder builder =
+                        UriComponentsBuilder.fromUriString(request.getServletPath());
+                if (builder != null
+                        && builder.build() != null
+                        && builder.build().getPath() != null) {
+                    Service serviceDescriptor =
+                            dispatcher.findService(
+                                    Objects.requireNonNull(builder.build().getPath()),
+                                    req.getVersion(),
+                                    req.getNamespace());
+                    if (serviceDescriptor != null) {
+                        req.setServiceDescriptor(serviceDescriptor);
+                        try {
+                            return dispatcher.fireServiceDispatchedCallback(req, serviceDescriptor);
+                        } catch (ServiceException se) {
+                            LOGGER.log(
+                                    Level.FINE,
+                                    "Exception while searching for the 'Service Descriptor'",
+                                    se);
+                            return null;
+                        }
```

Comment:
Why go through the callbacks, and not just return the service found?

Reply:
done